R/xp_utils.R

```diff
@@ -43,7 +44,13 @@ xml_nodes_to_lints <- function(xml, source_expression, lint_message,
     return(list())
   }
   if (inherits(xml, "xml_nodeset")) {
-    lints <- lapply(xml, xml_nodes_to_lints, source_expression, lint_message, type, match_after_end)
+    if (is.character(message) && length(message) > 1L) {
+      lints <- .mapply(xml_nodes_to_lints, list(xml = xml, lint_message = lint_message), list(
+        source_expression = source_expression, type = type, match_after_end = match_after_end
+      ))
+    } else {
+      lints <- lapply(xml, xml_nodes_to_lints, source_expression, lint_message, type, match_after_end)
+    }
```

Comment:
Not sure about the most elegant way. A simple `mapply()` will likely choke on `source_expressions` being a list.

Reply:
Seemed to do OK:

```
mapply(c, as.list(1:10), as.list(10:1), SIMPLIFY=FALSE)
mapply(c, replicate(10, sample(5, sample(5, 1)), simplify = FALSE), as.list(10:1), SIMPLIFY=FALSE)
mapply(c, replicate(10, as.list(sample(5, sample(5, 1))), simplify = FALSE), as.list(10:1), SIMPLIFY=FALSE)
```

Reply:
What does
```
mapply(list, xml = 1:10, source_expression = list(1:6), message = paste("m", 1:10))
```
do?

Reply:
throws a warning because of 6 < 10, but it gets recycled and then elt `i` of output is `list(xml[[ii]], source_expression[[ii %% 10]], message[[ii]])` (when `SIMPLIFY=FALSE`, then `simplify2array` of this otherwise)

Reply:
Yeah, but what we want is `source_expressions` to stay the same for every call, i.e. be recycled as length-1.
I'll test out if the `.mapply()` branch also works for the `length(lint_message) == 1L` case.

Reply:
No, it doesn't because if `lint_message` is a function, `.mapply()` stops with "object of type closure is not subsettable".

Reply:
the issue is `source_expression` has 6 elements and `mapply` will try and iterate over them right? won't `source_expression = list(source_expression)` solve that?

Reply:
c.f.

```
l = as.list(1:6)
x = 1:10
y = 10:1
mapply(c, x, list(l), y, SIMPLIFY=FALSE)
```

Reply:
Yeah, but it doesn't really help in any way compared to just passing it via `MoreArgs`, which is essentially what happens with the `.mapply()` call I wrote.